protocols/cosi/public_api.go

```diff
@@ -0,0 +1,459 @@
+	}
+	pcosi := node.ProtocolInstance().(*ProtocolCosi)
+	pcosi.SigningMessage(cr.Message)
+	hash := network.Suite.Hash()
+	if _, err := hash.Write(cr.Message); err != nil {
+		dbg.Error("Couldn't hash message:", err)
+	}
+	sum := hash.Sum(nil)
+	// Register the handler when the signature is finished
+	rchan := make(chan *SignResponse)
+	fn := func(chal, resp abstract.Secret) {
+		response := &SignResponse{
+			Sum:       sum,
+			Challenge: chal,
+			Response:  resp,
+		}
+		dbg.Lvl2("Getting CoSi signature back => sending to client")
+		// send back to client
+		rchan <- response
+	}
+	pcosi.RegisterDoneCallback(fn)
+	dbg.Lvl2("Starting CoSi protocol...")
+	go node.StartProtocol()
+	return <-rchan
+}
+
+// sign takes a stream and a toml file defining the servers
+func Sign(r io.Reader, tomlFileName string) (*SignResponse, error) {
+	dbg.Lvl3("Starting signature")
+	f, err := os.Open(tomlFileName)
+	if err != nil {
+		return nil, err
+	}
+	el, err := ReadGroupToml(f)
+	if err != nil {
+		return nil, err
+	}
+	dbg.Lvl2("Sending signature to", el)
+	res, err := SignStatement(r, el)
+	if err != nil {
+		return nil, err
+	}
+	return res, nil
+}
+
+// signStatement can be used to sign the contents passed in the io.Reader
+// (pass an io.File or use an strings.NewReader for strings)
+func SignStatement(r io.Reader, el *sda.EntityList) (*SignResponse, error) {
+
+	// create a throw-away key pair:
+	kp := config.NewKeyPair(network.Suite)
+
+	// create a throw-away entity with an empty  address:
+	e := network.NewEntity(kp.Public, "")
+	client := network.NewSecureTCPHost(kp.Secret, e)
+	msg, _ := crypto.HashStream(network.Suite.Hash(), r)
+	req := &SignRequest{
+		EntityList: el,
+		Message:    msg,
+	}
+
+	// Connect to the root
+	host := el.List[0]
+	dbg.Lvl3("Opening connection to", host.Addresses[0], host.Public)
+	con, err := client.Open(host)
+	defer client.Close()
+	if err != nil {
+		return nil, err
+	}
+
+	dbg.Lvl3("Sending sign SignRequest")
+	pchan := make(chan SignResponse)
+	go func() {
+		// send the SignRequest
+		if err := con.Send(context.TODO(), req); err != nil {
+			close(pchan)
+			return
+		}
+		dbg.Lvl3("Waiting for the response")
+		// wait for the response
+		packet, err := con.Receive(context.TODO())
+		if err != nil {
+			close(pchan)
+			return
+		}
+		pchan <- packet.Msg.(SignResponse)
+	}()
+	select {
+	case response, ok := <-pchan:
+		dbg.Lvl5("Response:", ok, response)
+		if !ok {
+			return nil, errors.New("Invalid repsonse: Could not cast the " +
+				"received response to the right type")
+		}
+		err = cosi.VerifySignature(network.Suite, msg, el.Aggregate,
+			response.Challenge, response.Response)
+		if err != nil {
+			return nil, err
+		}
+		return &response, nil
+	case <-time.After(time.Second * 10):
+		return nil, errors.New("Timeout on signing")
+	}
+}
+
+// verify takes a file and a group-definition, calls the signature
+// verification and prints the result
+func Verify(fileName, groupToml string) error {
```

Comment:
Could you put that directly into lib/cosi ? As there is no network communication nothing complex, it can be put directly in lib/cosi I think, and it's only a wrapper around the existing method that exists to verify (`cosi.VerifySignature`...)
PS: only doing high level comments since it's a very big PR.